### PR TITLE
CLI Improvements

### PR DESCRIPTION
**Pull request by lawrencelomax**
- Trim out garbage characters to improve parsing in interactive mode
- JSON Representable `FBWritableLog`
- `OptionSetType` union parser
- `--json` Flag
- JSON output for `diagnose`
- `Query` & `Format` optional, to be replaced with defaulted values.
- `Defaults` is now a thing and uses `Defaultable` values or reads from a `.rc` file
- Chaining of `Actions` is replaced with chaining of `Interactions`. This makes parsing less ambiguous and means that the 'implicit last query on chained interaction' dissapears, use `&&` in a shell to chain on different queries.
- Clarify `Interactions` 'Interactive Mode'
- Add `create` `Action` for explicitly creating based on a configuration.
- Add `delete` `Interacton` for deleting Simulators based on a query.

### CODE REVIEW

File: fbsimctl/FBSimulatorControlKit/Sources/Command.swift [hunk 1]
```diff
@@ -14,8 +14,18 @@ import FBSimulatorControl
...
 */
 public struct Configuration {
+  public struct Options : OptionSetType {
+    public let rawValue : Int
+    public init(rawValue: Int) {
+      self.rawValue = rawValue
+    }
+
+    static let DebugLogging  = Options(rawValue: 1 << 0)
+    static let JSONOutput  = Options(rawValue: 1 << 1)
```
Comment (asm89) [inline fbsimctl/FBSimulatorControlKit/Sources/Command.swift]: nit: double space intentional here?

File: fbsimctl/FBSimulatorControlKit/Sources/CommandParsers.swift [hunk 2]
```diff
@@ -72,6 +72,36 @@ extension Parser {
...
+        self.optionsParser(),
+        self.controlConfigurationParser()
+      )
+      .fmap { (options, controlConfiguration) in
+        return Configuration(controlConfiguration: controlConfiguration, options: options)
+      }
+  }
+
+  public static func controlConfigurationParser() -> Parser<FBSimulatorControlConfiguration> {
+    return Parser.ofTwoSequenced(
```
Comment (asm89) [inline fbsimctl/FBSimulatorControlKit/Sources/CommandParsers.swift]: nit: `.ofTwoSequenced(` on newline or line 78 on the previous line? :)

File: fbsimctl/FBSimulatorControlKit/Sources/Command.swift (hunk 1 above)
Comment (lawrencelomax) [inline fbsimctl/FBSimulatorControlKit/Sources/Command.swift]: Nope!

File: fbsimctl/FBSimulatorControlKit/Sources/CommandParsers.swift (hunk 2 above)
Comment (lawrencelomax) [inline fbsimctl/FBSimulatorControlKit/Sources/CommandParsers.swift]: I'm pulling this styling from 'what looks ok', but I think reading downwards vertically with the indent level signifying the nesting of parsers, makes the most sense so i'll move `ofTwoSequenced` onto a new line

Comment (asm89) [inline fbsimctl/FBSimulatorControlKit/Sources/CommandParsers.swift]: It was in a few other places as well. E.g. `Parser<FBSimulatorManagementOptions>.unionOptions([
` and `Parser.ofTwo...`. Not sure if you want it to be consistent across the codebase.

**Comment by asm89** [comment]
Apart from nits this LGTM.

**Pull request merged**